FileAudioSource/FileSourceAudioDevice.cs

```diff
@@ -0,0 +1,286 @@
+﻿using System;
+using System.IO;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace OpenTok
+{
+    public class FileSourceAudioDevice : IAudioDevice, IDisposable
+    {
+        #region Public
+
+        public void SelectSourceFile(string file)
+        {
+            if (!File.Exists(file))
+                throw new ArgumentException("File does not exist");
+            if (!file.EndsWith(".wav"))
+                throw new ArgumentException("Invalid file format");
+            lock (audioDeviceLock)
+            {
+                sourceFile = file;
+                AudioDevice.RestartInputAudioDevice();
+            }
+        }
+
+        public delegate void AudioPropertiesChangedHandler(int numberOfChannels, int samplingRate);
+        public event AudioPropertiesChangedHandler AudioPropertiesChanged;
+
+        #endregion
+
+        #region IAudioDevice
+
+        public void InitAudio(AudioDevice.AudioBus audioBus)
+        {
+            lock (audioDeviceLock)
+            {
+                this.audioBus = audioBus;
+            }
+        }
+
+        public void DestroyAudio()
+        {
+            lock (audioDeviceLock)
+            {
+                audioBus = null;
+            }
+        }
+
+        public void InitAudioCapturer()
+        {
+            lock (audioDeviceLock)
+            {
+                if (IsAudioCapturerInitialized())
+                    return;
+                if (!string.IsNullOrWhiteSpace(sourceFile))
+                {
+                    sourceFileStream = File.OpenRead(sourceFile);
+                    /* WAV file format header struct can be found at https://docs.fileformat.com/audio/wav/ */
+                    byte[] wavHeader = new byte[40];
+                    sourceFileStream.Read(wavHeader, 0, wavHeader.Length);
+                    capturerSettings = new AudioDeviceSettings();
+                    capturerSettings.NumChannels = wavHeader[22] | (wavHeader[23] << 8);
```

Comment:
Out of curiosity, no library to do this? In any case, it is simple enough.

Reply:
Yes I think there was one and it actually reads things like MP3 and so on. The thing is since the important part here is our SDK (and how to use it for custom audio devices) and not the actual file audio parsing method I preferred not using any third party libraries.